serialize.go

```diff
@@ -0,0 +1,153 @@
+type Serialized struct {
+	schema         string
+	data           []byte
+	sqliteOwnsData bool
+	shouldFreeData bool
+}
+
+// NewSerialized creates a new serialized DB from the given schema and data.
+//
+// If copyToSqlite is true, the data will be copied. This should be set to true
+// if this will be used with SQLITE_DESERIALIZE_FREEONCLOSE.
+func NewSerialized(schema string, data []byte, copyToSqlite bool) *Serialized {
+	s := &Serialized{
+		schema: schema,
+		data:   data,
+	}
+	if copyToSqlite {
+		sqliteData := (*[1 << 28]uint8)(unsafe.Pointer(C.sqlite3_malloc(C.int(len(data)))))[:len(data):len(data)]
+		copy(sqliteData, data)
+		s.data = sqliteData
+		s.shouldFreeData = true
+		s.sqliteOwnsData = true
+		runtime.SetFinalizer(s, func(s *Serialized) { s.free() })
+	}
+	return s
+}
+
+// Schema returns the schema for this serialized DB.
+func (s *Serialized) Schema() string {
+	if s.schema == "" {
+		return "main"
+	}
+	return s.schema
+}
+
+// Bytes returns the serialized bytes. Do not mutate this value. This is only
+// valid for the life of its receiver and should be copied for any other
+// longer-term use.
+func (s *Serialized) Bytes() []byte { return s.data }
+
+func (s *Serialized) free() {
+	if len(s.data) > 0 && s.shouldFreeData {
+		s.shouldFreeData = false
+		s.sqliteOwnsData = false
+		C.sqlite3_free(unsafe.Pointer(&s.data[0]))
+	}
+	s.data = nil
+}
+
+// SerializeFlags are flags used for Serialize.
+type SerializeFlags int
+
+const (
+	SQLITE_SERIALIZE_NOCOPY SerializeFlags = C.SQLITE_SERIALIZE_NOCOPY
+)
+
+// Serialize serializes the given schema. Returns nil on error. If
+// SQLITE_SERIALIZE_NOCOPY flag is set, the data may only be valid as long as
+// the database.
+//
+// https://www.sqlite.org/c3ref/serialize.html
+func (conn *Conn) Serialize(schema string, flags ...SerializeFlags) *Serialized {
+	var serializeFlags SerializeFlags
+	for _, f := range flags {
+		serializeFlags |= f
+	}
+
+	cschema := cmain
+	if schema != "" && schema != "main" {
+		cschema = C.CString(schema)
+		defer C.free(unsafe.Pointer(cschema))
+	}
+
+	var csize C.sqlite3_int64
+	res := C.sqlite3_serialize(conn.conn, cschema, &csize, C.uint(serializeFlags))
+	if res == nil {
+		return nil
+	}
+
+	s := &Serialized{
+		schema:         schema,
+		data:           (*[1 << 28]uint8)(unsafe.Pointer(res))[:csize:csize],
+		sqliteOwnsData: true,
+	}
+	// Free the memory only if they didn't specify nocopy
+	if serializeFlags&SQLITE_SERIALIZE_NOCOPY == 0 {
+		s.shouldFreeData = true
+		runtime.SetFinalizer(s, func(s *Serialized) { s.free() })
+	}
+	return s
+}
+
+// DeserializeFlags are flags used for Deserialize.
+type DeserializeFlags int
+
+const (
+	SQLITE_DESERIALIZE_FREEONCLOSE DeserializeFlags = C.SQLITE_DESERIALIZE_FREEONCLOSE
+	SQLITE_DESERIALIZE_RESIZEABLE  DeserializeFlags = C.SQLITE_DESERIALIZE_RESIZEABLE
+	SQLITE_DESERIALIZE_READONLY    DeserializeFlags = C.SQLITE_DESERIALIZE_READONLY
+)
```

Comment:
I think this is too low level to expose in the Go API here. When we deserialize we are taking go memory with the serialized DB and malloc-ing a space of C memory for which sqlite can operate an in-memory database. I think the only option we should expose is READONLY and maybe RESIZEABLE. I think since we are managing the C memory, we should control FREEONCLOSE, which we should probably set, so we don't have to continue to manage that memory.